misk-aws/src/main/kotlin/misk/jobqueue/sqs/SqsJobConsumer.kt

```diff
@@ -63,73 +70,106 @@ internal class SqsJobConsumer @Inject internal constructor(
+
+            client.receiveMessage(receiveRequest).messages
+          }
+        }
+      } catch (e: ClientExecutionTimeoutException) {
+        log.info("timed out long polling for messages from ${queue.queueName}")
+        emptyList<Message>()
       }
 
-      val queueName = queue.name
+      return messages.map { SqsJob(queue.name, queues, metrics, it) }
+    }
 
-      metrics.sqsReceiveTime.record(receiveDuration.toMillis().toDouble(), queueName.value, queueName.value)
+    private fun receive(): List<Status> {
+      val messages = fetchMessages()
 
       if (messages.size == 0) {
-        return Status.NO_WORK
+        return listOf(Status.NO_WORK)
       }
 
-      val futures = messages.map { SqsJob(queueName, queues, metrics, it) }.map { message ->
-        dispatchThreadPool.submit {
-          metrics.jobsReceived.labels(queueName.value, queueName.value).inc()
+      val futures = handleMessages(messages)
+
+      return CompletableFuture.allOf(*futures.toTypedArray()).thenApply {
+        futures.map { it.join() }
+      }.join()
+    }
 
-          tracer.traceWithSpan("handle-job-${queueName.value}") { span ->
+    private fun handleMessages(messages: List<SqsJob>): List<CompletableFuture<Status>> {
+      return messages.map { message ->
+        CompletableFuture.supplyAsync(Supplier {
+          metrics.jobsReceived.labels(queue.queueName, queue.queueName).inc()
+
+          tracer.traceWithSpan("handle-job-${queue.queueName}") { span ->
             // If the incoming job has an original trace id, set that as a tag on the new span.
             // We don't turn that into the parent of the current span because that would
             // incorrectly include the execution time of the job in the execution time of the
-            // action that triggered the job        q
+            // action that triggered the job
             message.attributes[SqsJob.ORIGINAL_TRACE_ID_ATTR]?.let {
               ORIGINAL_TRACE_ID_TAG.set(span, it)
             }
 
             // Run the handler and record timing
             try {
               val (duration, _) = timed { handler.handleJob(message) }
-              metrics.handlerDispatchTime.record(duration.toMillis().toDouble(), queueName.value, queueName.value)
+              metrics.handlerDispatchTime.record(duration.toMillis().toDouble(), queue.queueName,
+                  queue.queueName)
+              Status.OK
             } catch (th: Throwable) {
-              log.error(th) { "error handling job from ${queueName.value}" }
-              metrics.handlerFailures.labels(queueName.value, queueName.value).inc()
+              log.error(th) { "error handling job from ${queue.queueName}" }
+              metrics.handlerFailures.labels(queue.queueName, queue.queueName).inc()
               Tags.ERROR.set(span, true)
-              throw th
+              Status.FAILED
             }
           }
-        }
+        }, dispatchThreadPool)
```

Comment:
aren't these the consumers, i.e.: the threads that receivers have available to process messages? if so, I would just change the name, otherwise it starts getting super confusing.